luigi/contrib/external_program.py

```diff
@@ -132,6 +140,47 @@ def run(self):
                 tmp_stdout.close()
 
 
+class TrackingUrlContext(object):
+    time_to_sleep = 0.5
+
+    def __init__(self, main_proc_args, main_proc_kwargs, pattern, set_tracking_url):
+        # replace process output with PIPE for scanning
+        self.file_to_write = main_proc_kwargs.get('stdout')
+        main_proc_kwargs.update({'stdout': subprocess.PIPE})
+        self.main_proc = subprocess.Popen(main_proc_args, **main_proc_kwargs)
+        self.pipe_to_read = self.main_proc.stdout
+
+        self.pattern = pattern
+        self.set_tracking_url = set_tracking_url
+        self.track_proc = Process(target=self._track_url_by_pattern)
+
+    def __enter__(self):
```

Comment:
Is it possible to use contextlib? This reminds me of #2652. :)

Reply:
Thanks, done, now looks cleaner